docs/design/proposals/late-initialization/late-initialization.md

```diff
@@ -0,0 +1,552 @@
+    > For the resources with Async creation, the late initalized fields are not readily available after create operation. Therefore the reconciler loop needs to be repeated for these resources until the late initalized fields are available.
+
+### ACK Code Generator Changes
+1. Fix the latent bug where Spec fields are not set from the Create and Update response. Once this bug is fixed the late initialized fields which are present in Create and Update response will be set.
+    ```go
+    // Current Code
+    // sdk.go.tpl    
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 false }}
+        ...
+    }
+    ```
+
+    ```go
+    // Proposed Code
+    // sdk.go.tpl
+    func (rm *resourceManager) sdkCreate(
+	ctx context.Context,
+	desired *resource,
+    ) (created *resource, err error) {
+        ...
+        // NOTE: The performSpecUpdate boolean param will be removed
+        {{ GoCodeSetCreateOutput .CRD "resp" "ko" 1 }}
+        ...
+    }
+    ```
+
+2. **Provide a new "LateInitializationConfig" struct inside generator config for AWS service teams to provide which spec fields will be late initialized.** These fields will be the ones which are late initialized by the server but not present in the Create and Update response.
+
+    These fields will be populated by making the readOne call. AWS service teams can provide the configuration to either late initialize these fields immidiately after Create/Update call or provide a delay after which these fields will be late initialized.
+    LateInitializationConfig will be added as part of FieldConfig. 
+
+    ```go
+        // LateInitializeConfig contains instructions for how to handle the
+        // retrieval and setting of server-side defaulted fields.
+        type LateInitializationConfig struct {
+            // DefaultImmediately indicates that if the user has not
+            // specified a value for the field, then directly after a
+            // successful call to Create/Update, the reconciler will call
+            // ReadOne to get the value of the field as defaulted by the
+            // AWS service, set the value of this field on the resource and
+            // patch the resource back to the Kubernetes API.
+            DefaultImmidiately bool `json:"default_immidiately"`
+            // DelaySeconds provides the number of seconds to wait to set
+            // late initialized fields after successful Create/Update call
+            // DelaySeconds is only used if DefaultImmidiately is false
+            // Default value for DelaySeconds is 5 seconds if DefaultImmidiately
+            // is false and DelaySeconds is absent.
+            DelaySeconds *int `json:"delay_seconds,omitempty"`
```

Comment:
Do we need a `DefaultImmediately` field if we have a `DelaySeconds` field? You could make the `DelaySeconds` Go type an `int` instead of an `*int` and the default zero value would essentially mean "default immediately", right?

Reply:
Good point. I didn't realize earlier that empty structs can be mentioned in yaml.

How about following?

```go
type LateInitializationConfig struct {
    DelaySeconds int `json:"delay_seconds,omitempty"`
}
```
and `generator.yaml` will have entry like
```yaml
#For defaulting immediately
late_initialization_config: {}
``` 
```yaml
# for Defaulting with delay
late_initialization_config:
    delay_seconds: 10
```